runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5770,19 +5775,27 @@ class INTERPRETER_CLASS
 		case J9VM_CONTINUATION_RETURN_FROM_MONITOR_ENTER:
 			break;
 		case J9VM_CONTINUATION_RETURN_FROM_OBJECT_WAIT: {
-			j9object_t waitObject = *(j9object_t *)(_sp + 3);
-			rc = tryEnterBlockingMonitor(REGISTER_ARGS, waitObject, J9VM_CONTINUATION_RETURN_FROM_OBJECT_WAIT);
+			rc = tryEnterBlockingMonitor(REGISTER_ARGS, syncObject, J9VM_CONTINUATION_RETURN_FROM_OBJECT_WAIT);
 			if ((NULL != _currentThread->currentContinuation) && (EXECUTE_BYTECODE == rc)) {
-				waitObject = *(j9object_t *)(_sp + 3);
-				omrthread_monitor_t monitor = getMonitorForWait(_currentThread, waitObject);
+				syncObject = *(j9object_t *)(_sp + 3);
+				omrthread_monitor_t monitor = getMonitorForWait(_currentThread, syncObject);
 				monitor->count = _currentThread->currentContinuation->waitingMonitorEnterCount;
 				_currentThread->ownedMonitorCount += monitor->count - 1;
 				_currentThread->currentContinuation->waitingMonitorEnterCount = 0;
 
+				bool interrupted = J9VMJAVALANGTHREAD_DEADINTERRUPT(_currentThread, _currentThread->threadObject);
+				bool notified = J9VMJAVALANGVIRTUALTHREAD_NOTIFIED(_currentThread, _currentThread->threadObject);
+
+				if (J9_EVENT_IS_HOOKED(_vm->hookInterface, J9HOOK_VM_MONITOR_WAITED)) {
+					IDATA rc = (interrupted ? J9THREAD_INTERRUPTED : (notified ? 0 : J9THREAD_TIMED_OUT));
+					UDATA millis = J9VMJAVALANGVIRTUALTHREAD_TIMEOUT(_currentThread, _currentThread->threadObject);
+
+					/* Dispatch MonitorWaited hook, with stored metadata. */
+					TRIGGER_J9HOOK_VM_MONITOR_WAITED(_vm->hookInterface, _currentThread, monitor, millis, 0, rc, _currentThread->currentContinuation->startTicks, (UDATA) monitor, J9OBJECT_CLAZZ(_currentThread, syncObject));
```

Comment:
the parameter after `millis` is for `nanos`. would passing `0` for `nanos` lead to incorrect behaviour?
```suggestion
					TRIGGER_J9HOOK_VM_MONITOR_WAITED(_vm->hookInterface, _currentThread, monitor, millis, nanos, rc, _currentThread->currentContinuation->startTicks, (UDATA) monitor, J9OBJECT_CLAZZ(_currentThread, syncObject));
```

Reply:
it isn't here yet ... implication was not do we need `nanos` here?

Reply:
so the timeout used by JCL https://github.com/ibmruntimes/openj9-openjdk-jdk24/blob/openj9/src/java.base/share/classes/java/lang/VirtualThread.java#L630 is the value which we set in `BytecodeInterpreter.hpp::inlObjectWaitImpl` via `J9VMJAVALANGVIRTUALTHREAD_SET_TIMEOUT`

so to maintain consistency, Im using the save millisecond value here.
nanos field is required by the hook, which is why I put in a 0. if we need the nanos field, then we have to add additional metadata fields into either Continuation object or J9VMContinuation to store them.